src/main/java/one/util/streamex/StreamEx.java

```diff
@@ -299,6 +315,38 @@
         return new StreamEx<>(new PairSpliterator.PSOfRef<>(lastMapper, notLastMapper, spliterator(), false), context);
     }
 
+    /**
+     * Returns a stream where only elements that are instances of the provided {@code clazz} are transformed using
+     * {@code mapper} and other elements are left unmodified.
+     *
+     * <p>
+     *
+     * Examples:
+     * <p>
+     *   List[Object] objs = StreamEx.of(1, "a") <br>
+     *      .mapInstance(Integer.class, num -&gt; num * 2) <br>
+     *      .mapInstance(String.class, str -&gt; str + "_updated") <br>
+     *      .toList();
+     * <p>
+     *   List[Integer] ints = StreamEx.of(1, "2") <br>
+     *      .mapInstance(String.class, Integer::parseInt) <br>
+     *      .select(Integer.class)
+     *      .toList(); <br>
+     *
+     * @param clazz The class whose instances we want to map
+     * @param instanceMapper The mapping function for the elements that are instances of the provided class
+     * @param <R> The element type of the new stream
+     * @return The new stream
+     */
+    public <R extends T> StreamEx<Object> mapInstance(Class<R> clazz, Function<R, ?> instanceMapper) {
```

Comment:
I don't like this method signature at all. Why not `public <TT> StreamEx<T> mapInstance(Class<R> clazz, Function<? super R, ? extends T> instanceMapper)`? So we leave the stream object type the same as original, not `Object`, requiring the function to produce the original type. Your samples would work.

Reply:
I had it that way, but one of my tests won't compile with that signature:

```
        class A {}
        class B extends A {}
        class C extends A {}

        List<A> original = asList(new A(), new B(), new C());
        List<String> expected = asList("a", "b", "c");

        List<String> actual = StreamEx.of(original)
                .mapInstance(B.class, ignored -> "b")
                .mapInstance(C.class, ignored -> "c")
                .selectOrDefault(String.class, "a")
                .toList();

        assertEquals(expected, actual);
```

because after the first call `mapInstance(B.class, ignored -> "b")` you have a stream with mixed `A.class` and `String.class`, but with type `StreamEx<String>`.

You would have to coerce the stream to be of type object at creation time:
```
List<String> actual = StreamEx.<Object>of(original)
    ....
````

Reply:
@gm2211 That's too hacky. After the first `mapInstance`, you lose all type safety and can call `.mapInstance(Whatever.class, ....)`. You'd need to switch to some `ExStreamEx<A, String>` keeping both type arguments. And Java type inference most probably won't grasp it, anyway.

I guess, the nicest solution would be a builder for the mapping function in the very comment. You build it and then just `map` the stream.

Reply:
yeah - this PR is like 2 years old - at the time it seemed convenient but since then:

1. Java has evolved with better pattern-matching capabilities
2. Our projects have started using code-generated union types with staged-builders for visitors, which are way better at dealing with the original problem I had outlined